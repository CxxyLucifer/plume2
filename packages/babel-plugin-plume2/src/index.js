/**
 * 判断当前的call-expression是不是require('react-dom')
 * @param {*} t
 * @param {*} path
 */

function isRequireReactDOM(t, path) {
  return (
    t.isIdentifier(path.node.callee, { name: 'require' }) &&
    path.get('arguments.0').isStringLiteral({ value: 'react-dom' })
  );
}

module.exports = function(babel) {
  var t = babel.types;

  return {
    visitor: {
      /**
       * if current env is react-native
       * import ReactDOM from 'react-dom' => import ReactDOM from 'react-native'
       * @param {*} path
       * @param {*} param1
       */
      ImportDeclaration(path, { opts }) {
        if (
          opts.reactnative &&
          t.isStringLiteral(path.node.source, {
            value: 'react-dom'
          })
        ) {
          path.node.source = t.StringLiteral('react-native');
        }
      },

      /**
       * if current env is react-native
       * var ReactDOM = require('react-dom') => var ReactDOM = require('react-native');
       * @param {*} path
       * @param {*} param1
       */ CallExpression(path, { opts }) {
        if (opts.reactnative && isRequireReactDOM(t, path)) {
          path.node.arguments[0] = t.StringLiteral('react-native');
        }
      }
    }
  };
};
