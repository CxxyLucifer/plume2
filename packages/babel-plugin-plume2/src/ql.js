/**
 * 判断当前的callexpress是不是QL
 * @param {*} path 
 */
export const isQLCallExpress = path => {
  const { node } = path;
  return node.callee.name === 'QL';
};

/**
 * 是不是可用的QL的语法
 * @param {*} path 
 */
export const isValidQLSyntax = path => {
  const { node } = path;
};
