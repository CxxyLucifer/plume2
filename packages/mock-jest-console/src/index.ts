declare const jest: { fn: Function };

export class BaseMock {
  constructor() {
    this._logs = [];
    this.mock();
  }

  private _logs;

  mock() {}

  mockFn() {
    return jest.fn(log => this._logs.push(log));
  }

  get logs() {
    return this._logs;
  }

  reset() {
    this._logs = [];
  }
}

export class MockWarn extends BaseMock {
  mock() {
    console.warn = this.mockFn();
  }
}

export class MockLog extends BaseMock {
  mock() {
    console.log = this.mockFn();
  }
}

export class MockDir extends BaseMock {
  mock() {
    console.dir = this.mockFn();
  }
}

export class MockDebug extends BaseMock {
  mock() {
    console.debug = this.mockFn();
  }
}

export class MockError extends BaseMock {
  mock() {
    console.error = this.mockFn();
  }
}

export class MockTrace extends BaseMock {
  mock() {
    console.trace = this.mockFn();
  }
}

export class MockGroup extends BaseMock {
  mock() {
    console.group = this.mockFn();
  }
}

export class MockGroupCollapsed extends BaseMock {
  mock() {
    console.groupCollapsed = this.mockFn();
  }
}

export class MockTime extends BaseMock {
  mock() {
    console.time = this.mockFn();
  }
}

export class MockConsole extends BaseMock {
  mock() {
    console.warn = this.mockFn();
    console.log = this.mockFn();
    console.dir = this.mockFn();
    console.debug = this.mockFn();
    console.error = this.mockFn();
    console.trace = this.mockFn();
    console.group = this.mockFn();
    console.groupCollapsed = this.mockFn();
    console.time = this.mockFn();
  }
}
