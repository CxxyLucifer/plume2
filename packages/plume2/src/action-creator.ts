import StrEnum from 'str-enum';

export default StrEnum;
