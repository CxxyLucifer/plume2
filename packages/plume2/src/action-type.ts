import StrEnum from 'str-enum';

/**
 * 封装字符串的枚举类型
 * Usage:
 * actionType(
 *  'INIT',
 *  'LOADING_START'
 * )
 *
 * const Actions = actionType({
 *   PDF: 'application/pdf',
 *   Text: 'text/plain',
 *   JPEG: 'image/jpeg'
 * });
 */
export default StrEnum;
