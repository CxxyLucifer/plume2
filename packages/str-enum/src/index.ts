export default function StrEnum<T extends string>(
  ...args: Array<T>
): { [K in T]: K };

export default function StrEnum<
  T extends { [key: string]: V },
  V extends string
>(args: T): T;

export default function StrEnum(...args: Array<any>) {
  if (typeof args[0] === 'string') {
    const result = {} as any;
    for (let a of args) {
      result[a] = a;
    }
    return result;
  } else {
    return args[0];
  }
}
