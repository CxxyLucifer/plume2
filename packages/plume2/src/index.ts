import Actor from './actor';
import { Action } from './decorator';
import { go } from './go';
import msg from './msg';
import { PQL } from './pql';
import { QL } from './ql';
import Relax from './relax';
import Store from './store';
import StoreProvider from './store-provder';
import { IMap, IOptions, TViewAction } from './typing';
import { ViewAction } from './view-action';

export {
  go,
  QL,
  PQL,
  msg,
  Relax,
  ViewAction,
  Action,
  Actor,
  Store,
  StoreProvider,
  IMap,
  IOptions,
  TViewAction
};
