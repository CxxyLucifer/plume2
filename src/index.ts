import msg from './msg';
import { Action } from './decorator';
import Actor from './actor';
import Store from './store';
import StoreProvider from './store-provder';
import { QL } from './ql';
import Relax from './relax';

export { QL, msg, Relax, Action, Actor, Store, StoreProvider };
