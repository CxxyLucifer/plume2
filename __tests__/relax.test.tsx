import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Actor, Action, Store, StoreProvider, QL, Relax } from '../src/index';

//===============Actor=========================
class LoadingActor extends Actor {
  defaultState() {
    return {
      loading: false
    };
  }
}

class HelloActor extends Actor {
  defaultState() {
    return { mott: 'hello world!' };
  }

  @Action('change')
  change(state, text) {
    return state.set('mott', text);
  }
}

//===================Store======================
class AppStore extends Store {
  constructor(props) {
    super(props);
    window['_store'] = this;
  }

  bindActor() {
    return [new LoadingActor(), new HelloActor()];
  }
}

//========================UI======================
@StoreProvider(AppStore, { debug: false })
class HelloApp extends React.Component {
  render() {
    return <HelloRelax />;
  }
}

const loadingQL = QL('loadingQL', [
  'loading',
  /**
   * 
   */
  loading => loading
]);

const mottQL = QL('mottQL', [
  loadingQL,
  'mott',
  /**
   * 
   */
  (loading, mott) => ({ loading, mott })
]);

const loadingDQL = QL('loadingDQL', [
  '$mottFlag',
  /**
   * 
   */
  loading => loading
]);

@Relax
class HelloRelax extends React.Component {
  props: {
    mottFlag?: string;
    relaxProps?: {
      loading: boolean;
      mott: string;
      loadingQL: boolean;
      mottQL: { loading: boolean; mott: string };
      loadingDQL: boolean;
    };
  };

  static defaultProps = {
    mottFlag: 'mott'
  };

  static relaxProps = {
    loading: 'loading',
    mott: 'mott',
    loadingQL,
    mottQL,
    loadingDQL
  };

  render() {
    const {
      loading,
      mott,
      loadingQL,
      mottQL,
      loadingDQL
    } = this.props.relaxProps;

    expect(false).toEqual(loadingQL);
    expect({ loading: false, mott: 'hello world!' }).toEqual(mottQL);
    expect('hello world!').toEqual(loadingDQL);

    return (
      <div>
        <div>
          {loading}
        </div>
        <div>
          {mott}
        </div>
      </div>
    );
  }
}

describe('relax test suite', () => {
  it('initial render relax', () => {
    const tree = renderer.create(<HelloApp />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('dispatch event', () => {
    @StoreProvider(AppStore)
    class HelloApp extends React.Component {
      render() {
        return <Hello />;
      }
    }

    @Relax
    class Hello extends React.Component {
      props: { relaxProps?: { mott: string } };

      static relaxProps = {
        mott: 'mott'
      };

      render() {
        return (
          <div>
            <div>
              {this.props.relaxProps.mott}
            </div>
          </div>
        );
      }
    }

    const component = renderer.create(<HelloApp />);
    const store = window['_store'] as AppStore;
    store.dispatch('change', 'hello plume');
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('no relax-props warnning', () => {
    @StoreProvider(AppStore)
    class WarnApp extends React.Component {
      render() {
        return <RelaxTest />;
      }
    }

    @Relax
    class RelaxTest extends React.Component {
      render() {
        return <div />;
      }
    }

    renderer.create(<WarnApp />).toJSON();
  });
});
