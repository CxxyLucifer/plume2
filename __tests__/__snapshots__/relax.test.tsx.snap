// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relax test suite dispatch event 1`] = `
<div>
  <div>
    hello plume
  </div>
</div>
`;

exports[`relax test suite initial render relax 1`] = `
<div>
  <div />
  <div>
    hello world!
  </div>
</div>
`;
